File: src/main/java/com/yupi/airouter/service/BlacklistService.java
/**
 * 黑名单服务接口
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
package com.yupi.airouter.service;

import java.util.Set;

/**
 * 黑名单服务
 * 提供 IP 黑名单管理能力
 */
public interface BlacklistService {

    /**
     * 检查 IP 是否在黑名单中
     *
     * @param ip IP 地址
     * @return true 表示在黑名单中
     */
    boolean isBlocked(String ip);

    /**
     * 将 IP 加入黑名单
     *
     * @param ip     IP 地址
     * @param reason 封禁原因
     */
    void addToBlacklist(String ip, String reason);

    /**
     * 将 IP 从黑名单移除
     *
     * @param ip IP 地址
     */
    void removeFromBlacklist(String ip);

    /**
     * 获取所有黑名单 IP
     *
     * @return 黑名单 IP 集合
     */
    Set<String> getAllBlacklist();

    /**
     * 清空黑名单
     */
    void clearBlacklist();

    /**
     * 获取黑名单数量
     *
     * @return 黑名单数量
     */
    long getBlacklistCount();
}

File: python-backend/app/services/blacklist_service.py
"""IP blacklist service based on Redis Set."""

from __future__ import annotations

import logging

from redis.asyncio import Redis

from app.core.constants import BLACKLIST_IP_KEY

logger = logging.getLogger("app")


class BlacklistService:
    def __init__(self, redis: Redis) -> None:
        self.redis = redis

    async def is_blocked(self, ip: str | None) -> bool:
        if not ip or not ip.strip():
            return False
        return bool(await self.redis.sismember(BLACKLIST_IP_KEY, ip))

    async def add_to_blacklist(self, ip: str | None, reason: str | None = None) -> None:
        if not ip or not ip.strip():
            return
        await self.redis.sadd(BLACKLIST_IP_KEY, ip)
        logger.info("IP added to blacklist: %s, reason: %s", ip, reason)

    async def remove_from_blacklist(self, ip: str | None) -> None:
        if not ip or not ip.strip():
            return
        await self.redis.srem(BLACKLIST_IP_KEY, ip)
        logger.info("IP removed from blacklist: %s", ip)

    async def get_all_blacklist(self) -> set[str]:
        members = await self.redis.smembers(BLACKLIST_IP_KEY)
        return set(members)

    async def clear_blacklist(self) -> None:
        await self.redis.delete(BLACKLIST_IP_KEY)
        logger.info("Blacklist cleared")

    async def get_blacklist_count(self) -> int:
        return int(await self.redis.scard(BLACKLIST_IP_KEY))
